src/lib/y2network/widgets/bond_slave.rb

```diff
@@ -0,0 +1,217 @@
+require "cwm/custom_widget"
+require "y2network/widgets/slave_items"
+
+Yast.import "UI"
+Yast.import "LanItems"
+
+module Y2Network
+  module Widgets
+    class BondSlave < CWM::CustomWidget
+      include SlaveItems
+
+      def initialize(settings)
+        textdomain "network"
+        @settings = settings
+      end
+
+      def contents
+        Frame(
+          _("Bond Slaves and Order"),
+          VBox(
+            MultiSelectionBox(Id(:bond_slaves_items), Opt(:notify), "", []),
+            HBox(
+              PushButton(Id(:bond_slaves_up), Opt(:disabled), _("Up")),
+              PushButton(Id(:bond_slaves_down), Opt(:disabled), _("Down"))
+            )
+          )
+        )
+      end
+
+      def handle(event)
+        if event["EventReason"] == "SelectionChanged"
+          enable_slave_buttons
+        elsif event["EventReason"] == "Activated" && event["WidgetClass"] == :PushButton
+          items = ui_items || []
+          current = value.to_s
+          index = value_index
+          case event["ID"]
+          when :up
+            items[index], items[index - 1] = items[index - 1], items[index]
+          when :down
+            items[index], items[index + 1] = items[index + 1], items[index]
+          else
+            log.warn("unknown action #{event["ID"]}")
+            return nil
+          end
+          UI.ChangeWidget(:bond_slaves_items, :Items, items)
+          UI.ChangeWidget(:bond_slaves_items, :CurrentItem, current)
+          enable_slave_buttons
+        else
+          log.debug("event:#{event}")
+        end
+
+        nil
+      end
+
+      def help
+        # TODO: write it
+        ""
+      end
+
+      # Default function to init the value of slave devices box for bonding.
+      def init
+        # TODO: why? it should respect previous @settings, not?
+        @settings["SLAVES"] = Yast::LanItems.bond_slaves || []
+
+        @settings["BONDOPTION"] = Yast::LanItems.bond_option
+
+        items = slave_items_from(
+          Yast::LanItems.GetBondableInterfaces(Yast::LanItems.GetCurrentName),
+          Yast::LanItems.bond_slaves
+        )
+
+        # reorder the items
+        l1, l2 = items.partition { |t| @settings["SLAVES"].include? t[0][0] }
+
+        items = l1 + l2.sort_by { |t| justify_dev_name(t[0][0]) }
+
+        Yast::UI.ChangeWidget(:bond_slaves_items, :Items, items)
+
+        Yast::UI.ChangeWidget(
+          :bond_slaves_items,
+          :SelectedItems,
+          @settings["SLAVES"]
+        )
+
+        enable_slave_buttons
+
+        nil
+      end
+
+      # Default function to store the value of slave devices box.
+      def store
+        configured_slaves = @settings["SLAVES"] || []
+
+        selected_slaves = Yast::UI.QueryWidget(:bond_slaves_items, :SelectedItems) || []
+
+        @settings["SLAVES"] = selected_slaves
+
+        # XXX: Hmm, it stores different widget? This is dark cwm hack and hidden dependency
+        # that both widget have to be used together
```

Comment:
Well, it is a CustomWidget with a ComboBox and a Custom MultiSelectionBox with order, so both can store their state separetly and are not dependend each other although can be seen like a Bonding widget which will initialize and store both widget. so it is so black or dark cwm hack after all?

Reply:
@teclator problem is that BONDOPTION widget does not have store and init and instead it is done in BONDSLAVES widget. So it works only when both CWM widgets is used together, so completely against CWM idea of reusable widgets.

Reply:
@jreidinger mmm revising the code you are right which means the bond_options widget should be added, or it should be a `Bonding` widget with both.